server/src/fstar_connection.ts

```diff
@@ -0,0 +1,512 @@
+	// messages interface definition. Should also be documented in
+	// https://github.com/FStarLang/FStar/wiki/Editor-support-for-F*#cancel
+	async cancelRequest(range: { line: number; character: number }, debug: boolean) {
+		const query: CancelRequest = {
+			query: "cancel",
+			args: {
+				"cancel-line": range.line + 1,
+				"cancel-column": range.character
+			}
+		};
+
+		// TODO(klinvill): I believe there's no responses to cancel requests, is
+		// that correct?
+		const expect_response = false;
+		this.request(query, debug, expect_response);
+	}
+
+	async restartSolverRequest(debug: boolean) {
+		const query = {
+			query: "restart-solver",
+			args: {}
+		};
+
+		// TODO(klinvill): I believe there's no responses to restart-solver requests, is
+		// that correct?
+		const expect_response = false;
+		this.request(query, debug, expect_response);
+	}
+
+	////////////////////////////////////////////////////////////////////////////////////
+	// Methods to handle responses from F*
+	////////////////////////////////////////////////////////////////////////////////////
+
+	private handleResponse(msg: object) {
+		// Events for specific message types
+		if (isProtocolInfo(msg)) {
+			this.handleProtocolInfo(msg as ProtocolInfo);
+		} else {
+			// Either we expect unprompted protocol-info messages, or we expect
+			// responses to a query we sent.
+			const r = msg as IdeQueryResponse;
+			// TODO(klinvill): it appears that responses to full-buffer queries
+			// are sent with query ids with incrementing fractional components.
+			// E.g. if the query id is 2, the responses seem to come back with
+			// query ids 2, 2.1, 2.2, 2.3, etc. To deal with this behavior I
+			// truncate the responses query-id. Is this the proper behavior for
+			// this case?
+			const qid = Math.trunc(Number(r["query-id"]));
+			if (!qid) {
+				console.warn(`Ill-formed query response message: ${r}`);
+				return;
+			}
+
+			if (r.kind === 'message' && r.level === 'progress') {
+				const response = r.contents as IdeProgress;
+				// Progress responses are sent as a stream of messages. The last
+				// message is always "full-buffer-finished".
+				const done = response.stage === 'full-buffer-finished';
+				this.respondStream(qid, response, done);
+			} else if (r.kind === 'message' && r.level === 'proof-state') {
+				// TODO(klinvill): What queries can prompt proof-state status messages? Are they part of a stream (like responses to full-buffer queries)?
+				this.respond(qid, r.contents as IdeProofState);
+			} else if (r.kind === 'message' && r.level === 'info') {
+				// TODO(klinvill): info messages are just logged and don't resolve the corresponding pending_responses promise. Is this the right behavior (that info messages are just extraneous and should be logged and ignored)?
+				console.log("Info: " + r.contents);
+			} else if (r.kind === "response" && r.response) {
+				const responseType = FStarConnection.decideIdeReponseType(r.response);
```

Comment:
Is there a reason why we need to dispatch on the (heuristically guessed) response type here?  I think we do the same in all cases (i.e., resolve the promise associated with the query id).

Reply:
Ah that was the case, but with my latest change I wrap IdeError responses with a `kind` field that can be used to easily distinguish them (instead of heuristically guessing based on the response being an array or instead of passing through the full `IdeQueryResponse` object). Do you think there's a better way to do this?

Reply:
The proper way to figure out if a response was a success or not is by looking at the `status` field.  But I would keep it simple and just pass the whole `msg` back to the caller.  Then we can check `resp.status == 'error'` etc. at the call site.

Reply:
I wouldn't constrain the type parameters so much:
```typescript
// A query response envelope
export interface IdeQueryResponse {
	'query-id': string;
	kind: 'protocol-info' | 'response' | 'message';
	status?: 'success' | 'failure';
	level?: 'progress' | 'proof-state' | 'info';
}

export interface IdeQueryResponseR<R> extends IdeQueryResponse {
	response: R;
}
export interface IdeQueryResponseC<C> extends IdeQueryResponse {
	contents: C;
}
```

With the above, you'll even know that the `response` field of `IdeQueryResponseR` is defined.

Another maybe even clearer option is to have one type for every response:
```typescript
export interface IdeQueryResponse {
	'query-id': string;
	status?: 'success' | 'failure';
}

export interface IdeAutoCompleteResponse extends IdeQueryResponse {
	response: [number, string, string][];
}

// ...
```

For these kind of interface descriptions, I find inheritance to be a lot more readable than genericity.  (In case it's not clear, in typescript the value `{'query-id': '5', foo: 42}` does have the type `IdeQueryResponse`.  You can always have additional properties that are not mentioned in the interface.)

Reply:
Sounds good to me, I like the approach of having one type for every response. I went ahead and pushed up a commit that does this. The downside is there's now quite a few mostly trivial types (e.g. `IdeSymbolResponse` just extends `IdeQueryResponse` and adds the field `response: IdeSymbol`), but we can always refactor this later to move the details into the response type as desired (e.g. move the `IdeSymbol` definition into `IdeSymbolResponse`).

Reply:
I see what you mean.  I wouldn't worry too much about the details.  The important part is that the types check out.